drivers/freespace/sensor_api.cr

```diff
@@ -0,0 +1,243 @@
+  @password : String = ""
+  @connected : Bool = false
+
+  @floor_mappings : Hash(String, NamedTuple(building_id: String?, level_id: String)) = {} of String => NamedTuple(building_id: String?, level_id: String)
+  # Level zone => location_id
+  @zone_mappings : Hash(String, Array(String)) = Hash(String, Array(String)).new { |hash, key| hash[key] = [] of String }
+
+  def connected
+    @connected = true
+
+    # Send the CONNECT message
+    hostname = URI.parse(config.uri.not_nil!).hostname.not_nil!
+    @client = STOMP::Client.new(hostname)
+    send(client.stomp.to_s)
+
+    schedule.clear
+    schedule.in(5.seconds) { @auth_key = nil }
+    schedule.every(10.seconds) { heart_beat }
+  end
+
+  def disconnected
+    @connected = false
+    schedule.clear
+    @spaces.clear
+    @auth_key = @client = nil
+  end
+
+  def heart_beat
+    send(client.send("/beat/#{Time.utc.to_unix}").to_s, wait: false, priority: 0)
+  end
+
+  protected def subscribe_location(location_id) : Nil
+    get_location(location_id).each do |space|
+      id = space.id
+      request = client.subscribe("space-#{id}", "/topic/spaces/#{id}/activities", HTTP::Headers{
+        "receipt" => "rec-#{id}",
+      })
+
+      # Wait false as the server is not STOMP compliant, it won't respond to receipt headers
+      send(request.to_s, wait: false)
+    end
+  end
+
+  @[Security(Level::Support)]
+  def spaces_details
+    @spaces
+  end
+
+  @[Security(Level::Support)]
+  def spaces_state
+    @space_state
+  end
+
+  @[Security(Level::Support)]
+  def get_location(location_id : String | Int64) : Array(Space)
+    response = http("POST",
+      "/api/locations/#{location_id}/spaces",
+      headers: {
+        "X-Auth-Key"   => get_token,
+        "Content-Type" => "application/json",
+        "Accept"       => "application/json",
+      }, body: {
+        username: @username,
+        password: @password,
+      }.to_json
+    )
+
+    raise "issue obtaining to location #{location_id}: status code #{response.status_code}\n#{response.body}" unless response.success?
+
+    spaces = Array(Space).from_json response.body
+    spaces.each { |space| @spaces[space.id] = space }
+    spaces
+  end
+
+  # Alternative to using basic auth, but here really only for testing with postman
+  @[Security(Level::Support)]
+  def get_token : String
+    auth_key = @auth_key
+    return auth_key if auth_key
+
+    response = http("POST",
+      "/login",
+      headers: {
+        "Content-Type" => "application/json",
+        "Accept"       => "application/json",
+      }, body: {
+        username: @username,
+        password: @password,
+      }.to_json
+    )
+    logger.debug { "login response: #{response.body}" }
+    raise "issue obtaining token: #{response.status_code}\n#{response.body}" unless response.success?
+
+    # auth key is valid for 5 seconds
+    schedule.in(5.seconds) { @auth_key = nil }
+    @auth_key = response.headers["X-Auth-Key"]
+  end
+
+  def received(bytes, task)
```

Comment:
Should this be public?

Reply:
it's a protected method so not exposed to the public API and only used internally by `DriverManager`